blueocean-pipeline-scm-api/src/main/java/io/jenkins/blueocean/credential/UserSSHKeyManager.java

```diff
@@ -0,0 +1,169 @@
+import hudson.model.User;
+import hudson.remoting.Base64;
+import io.jenkins.blueocean.commons.ServiceException;
+import java.io.IOException;
+import java.security.KeyFactory;
+import java.security.KeyPair;
+import java.security.NoSuchAlgorithmException;
+import java.security.interfaces.RSAPrivateCrtKey;
+import java.security.interfaces.RSAPrivateKey;
+import java.security.interfaces.RSAPublicKey;
+import java.security.spec.InvalidKeySpecException;
+import java.security.spec.PKCS8EncodedKeySpec;
+import java.security.spec.RSAPublicKeySpec;
+import java.util.List;
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import jenkins.model.Jenkins;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+
+/**
+ * Helper to deal with managing a Jenkins-created per-user SSH key
+ * @author kzantow
+ */
+@Restricted(NoExternalUse.class)
+public class UserSSHKeyManager {
+    private static final int KEY_SIZE = 2048;
+    private static final String BLUEOCEAN_GENERATED_SSH_KEY_ID = "jenkins-generated-ssh-key";
+    private static final String BLUEOCEAN_DOMAIN_NAME = "blueocean-git-domain";
+    private static final Domain BLUEOCEAN_DOMAIN = new Domain(BLUEOCEAN_DOMAIN_NAME, null, null);
+    
+    /**
+     * Gets the existing generated SSH key for the user or creates one and
+     * returns it in the user's credential store
+     * @param user owner of the key
+     * @return the user's personal private key
+     */
+    public static @Nonnull BasicSSHUserPrivateKey getOrCreate(@Nonnull User user) {
+        Preconditions.checkNotNull(user);
+        
+        CredentialsStore store = getUserStore(user);
+        if(store == null){
+            throw new ServiceException.ForbiddenException(String.format("Logged in user: %s doesn't have writable credentials store", user.getId()));
+        }
+        // try to find the right key
+        for (Credentials cred : store.getCredentials(BLUEOCEAN_DOMAIN)) {
+            if (cred instanceof BasicSSHUserPrivateKey) {
+                BasicSSHUserPrivateKey sshKey = (BasicSSHUserPrivateKey)cred;
+                if (BLUEOCEAN_GENERATED_SSH_KEY_ID.equals(sshKey.getId())) {
+                    return sshKey;
+                }
+            }
+        }
+        // if none found, create one
+        BasicSSHUserPrivateKey key;
+        try {
+            // create one!
+            KeyPair keyPair = SSHKeyUtils.generateRSAKey(KEY_SIZE);
+            RSAPrivateKey privateKey = (RSAPrivateKey)keyPair.getPrivate();
+            String id_rsa = Base64.encode(privateKey.getEncoded());
+            BasicSSHUserPrivateKey.DirectEntryPrivateKeySource keySource = new BasicSSHUserPrivateKey.DirectEntryPrivateKeySource(id_rsa);
+            key = new BasicSSHUserPrivateKey(CredentialsScope.USER, BLUEOCEAN_GENERATED_SSH_KEY_ID, user.getId(), keySource, null, BLUEOCEAN_GENERATED_SSH_KEY_ID);
+            store.addCredentials(BLUEOCEAN_DOMAIN, key);
```

Comment:
Where is domain BLUEOCEAN_DOMAIN getting created? Look at [CredentialApi.create()](https://github.com/jenkinsci/blueocean-plugin/blob/a986f24023c4192f94ce6fc5b28b2f4b1e7b1626/blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/credential/CredentialApi.java#L55). but we should not create these APIs when we already have one.